### PR TITLE
kubeadm join the cluster with pre-existing client certificate if provided

### LINKED ISSUE #945: Support `kubeadm join` with a pre-existing client certificate
## Is this a request for help?

No.

## Is this a BUG REPORT or FEATURE REQUEST?

FEATURE REQUEST

## Detail

I'd like to use `kubeadm join` with a pre-existing client certificate (no token/TLSBootstrap).

We've got a environment where interacting with the master to setup a bootstrap token and feeding that to the new node is impractical/sub-optimal (security boundaries, provisioning latency, machine orchestration tooling, privileged credential practices, etc)

We currently generate our cluster CA & worker node client certificates externally and lay those down during system provisioning. (So far, it's been a pretty smooth approach)

On the master we run `kubeadm init`, which will detect and use our existing CA files.  On worker nodes, we lay down a suitable `/etc/kubernetes/kubelet.conf` with the externally generated node certificate. (This approach also allows us to add workers at a later date without having to interact with the master - our deployment tooling has the original CA keys and creates a suitable client-certificate externally)

I've noticed `kubeadm join` appears to be doing an increasing amount of "extrajoin" configuration (e.g. CRISocket, DynamicKubeletConfig) and imagine that this is will only increase in the future.

I feel that our manual `kubelet.conf` approach will become faulty or high maintenance if we must mimic a set of increasing requirements performed by `kubeadm join`.

Would it make sense to support:
a) Exposing `kubeadm join` phases via `kubeadm alpha phase` style commands
b) Use of existing on-disk `kubelet.conf` during `kubeadm join` (akin to CA files during `kubeadm init`).
c) Explicitly providomg node CA/client certificate "credentials" to `kubeadm join`

Or is our external certificate generation approach doomed for other reasons? (And we should start making other plans)

I hope I've explained that clearly.

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
support `kubeadm join` with a pre-existing client certificate

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#945

**Special notes for your reviewer**:
/cc @luxas @timothysc  @kubernetes/sig-cluster-lifecycle-pr-reviews 

**Release note**:

```release-note
kubeadm now can join the cluster with pre-existing client certificate if provided
```

### CODE REVIEW

File: cmd/kubeadm/app/discovery/discovery.go [hunk 1]
```diff
@@ -36,30 +36,34 @@ const TokenUser = "tls-bootstrap-token-user"
...
 	}
 
-	return kubeconfigutil.CreateWithToken(
-		clusterinfo.Server,
-		cfg.ClusterName,
-		TokenUser,
-		clusterinfo.CertificateAuthorityData,
-		cfg.TLSBootstrapToken,
-	), nil
+	if len(cfg.TLSBootstrapToken) != 0 {
```
Comment (neolit123) [inline cmd/kubeadm/app/discovery/discovery.go]:
```
if len(cfg.TLSBootstrapToken) == 0 {
	config, nil
}
...
```

to avoid indentation for the bellow block.

File: cmd/kubeadm/app/discovery/file/file.go [hunk 2]
```diff
@@ -116,14 +152,30 @@ func tryParseClusterInfoFromConfigMap(cm *v1.ConfigMap) (*clientcmdapi.Config, e
...
 	if defaultCluster == nil {
 		return fmt.Errorf("the provided cluster-info KubeConfig file must have an unnamed Cluster or a CurrentContext that specifies a non-nil Cluster")
 	}
-	return nil
+	return clientcmd.Validate(*config)
+}
+
+// extractFromFile tries to read the pem file
+func extractFromFile(filename string) ([]byte, error) {
+	glog.V(3).Infof("Reading config file %q", filename)
```
Comment (neolit123) [inline cmd/kubeadm/app/discovery/file/file.go]: don't we have the generic "open file" -> return `[]byte, error` function already?
if not, i think this should go in utils - e.g. `readFileBytes()`?

this would avoid the imports here too:
```
+	"io/ioutil"
+	"os"
+
+	"github.com/golang/glog"
```

File: cmd/kubeadm/app/discovery/file/file.go [hunk 3]
```diff
@@ -30,37 +34,69 @@ import (
...
+	} else {
+		configFromClusterInfo = kubeconfigutil.CreateBasic(
+			defaultCluster.Server,
+			clustername,
+			"", // no user provided
+			defaultCluster.CertificateAuthorityData,
+		)
+	}
 
 	client, err := kubeconfigutil.ToClientSet(configFromClusterInfo)
```
Comment (luxas) [inline cmd/kubeadm/app/discovery/file/file.go]: You should/can now remove this code that looks up the cluster info ConfigMap, not needed anymore. It could make sense to do that in a separate PR that would be merged first, but I don't have a strong preference here.

File: cmd/kubeadm/app/discovery/file/file.go [hunk 4]
```diff
@@ -30,37 +34,69 @@ import (
...
 	}
 
 	// This is the cluster object we've got from the cluster-info KubeConfig file
-	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(clusterinfo)
+	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
+
+	var authInfo *clientcmdapi.AuthInfo
+	if config.Contexts[config.CurrentContext] != nil {
+		// load pre-existing client certificates
+		authInfo = config.AuthInfos[config.Contexts[config.CurrentContext].AuthInfo]
```
Comment (luxas) [inline cmd/kubeadm/app/discovery/file/file.go]: After this, one more line and if block that checks for `authInfo != nil`

File: cmd/kubeadm/app/discovery/file/file.go [hunk 5]
```diff
@@ -30,37 +34,69 @@ import (
...
 	if err != nil {
 		return nil, err
 	}
 
 	// This is the cluster object we've got from the cluster-info KubeConfig file
-	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(clusterinfo)
+	defaultCluster := kubeconfigutil.GetClusterFromKubeConfig(config)
+
+	var authInfo *clientcmdapi.AuthInfo
+	if config.Contexts[config.CurrentContext] != nil {
```
Comment (luxas) [inline cmd/kubeadm/app/discovery/file/file.go]: and `len(config.AuthInfos) > 0`

File: cmd/kubeadm/app/discovery/file/file.go [hunk 6]
```diff
@@ -30,37 +34,69 @@ import (
...
+			}
+			authInfo.ClientCertificateData = clientCert
+		}
+		if len(authInfo.ClientKeyData) == 0 && len(authInfo.ClientKey) != 0 {
+			clientKey, err := extractFromFile(authInfo.ClientKey)
+			if err != nil {
+				return nil, err
+			}
+			authInfo.ClientKeyData = clientKey
+		}
```
Comment (luxas) [inline cmd/kubeadm/app/discovery/file/file.go]: You can just move this code over here and return what was found:
```
if len(authInfo.ClientCertificateData) == 0 || len(authInfo.ClientKeyData) == 0 {
    return nil, fmt.Errorf("couldn't read authentication info from the given kubeconfig file")
}
return configFromClusterInfo = kubeconfigutil.CreateWithCerts(...)
```

File: cmd/kubeadm/app/discovery/file/file.go [hunk 7]
```diff
@@ -30,37 +34,69 @@ import (
...
+		configFromClusterInfo = kubeconfigutil.CreateWithCerts(
+			defaultCluster.Server,
+			clustername,
+			"", // no user provided
+			defaultCluster.CertificateAuthorityData,
+			authInfo.ClientKeyData,
+			authInfo.ClientCertificateData,
+		)
+	} else {
+		configFromClusterInfo = kubeconfigutil.CreateBasic(
```
Comment (luxas) [inline cmd/kubeadm/app/discovery/file/file.go]: this can be left as-is, but just return it directly here.

**Review by luxas — COMMENTED** [review]
Great start :clap:!

**Review by timothysc — APPROVED** [review]
/approve 

Other can lgtm when ready.

**Comment by neolit123** [comment]
/lgtm
thank you @dixudx
